### ISSUE TITLE
NTLM Authentication Suddenly Stopped Working

### ISSUE THREAD

**Issue by Dangerunicorn** [a85/POSTMan-Chrome-Extension#7381]
Postman Application: Postman for Windows version 7.7.3

**Describe the bug**
NTLM Authentication suddenly stopped working for me. September 18th my suite of tests ran without issue, but when I ran them again yesterday (9/23) all the tests using NTLM are showing a 401 unauthorized error with the error "JSONError | No data, empty input at 1:1" appearing in the console as well as the developer tools. There have been no changes to any of my tests, or the setup of my machine between this period of time. No changes to usernames or passwords. I've tried using the credentials of other users and all have the same error. All my tests using other methods of authentication run without issue. 

**To Reproduce**
There isn't a good way to reproduce this issue. Unsure what makes it happen. 

**Expected behavior**
NTLM Authentication to work consistently. 

**Screenshots**
![Something Went Wrong](https://user-images.githubusercontent.com/54122108/65532137-22be0680-deb0-11e9-8ee0-23c6966acc72.PNG)

![JSONError](https://user-images.githubusercontent.com/54122108/65532148-294c7e00-deb0-11e9-9052-a0ef3d6f9ce8.PNG)

![Console](https://user-images.githubusercontent.com/54122108/65532661-16867900-deb1-11e9-8ac8-a2f4ba7cc1a8.PNG)

**App information (please complete the following information):**
 - App Type: Native App
 - Postman Version: 7.7.3
 - OS: Windows 10

**Comment by codenirvana** [comment]
@Dangerunicorn we need more information to reproduce this issue internaly.
By any chance is it possible this is due to an internal server error?

**Comment by Dangerunicorn** [comment]
Just tell me what info you need and I can provide it. This is not an internal server error. We've had no other issues. All other tests with other methods of authentication run fine. All of the endpoints I'm trying to hit can be accessed in browser without issue. 

Again, nothing has changed for us between those two dates. One day all the tests using NTLM passed, and a few days later they all fail. 

Edit: I see that Postman was updated on the 19th and again today the 25th. Could these updates be the culprit? The dates match up.

**Comment by Dangerunicorn** [comment]
What information can I provide for you?

Is this maybe related to issue #4355?

**Comment by codenirvana** [comment]
@Dangerunicorn Can you check if the request (just NTLM auth) works by removing the request body?

As you mentioned the request used to work previously and there's no change made in the NTLM auth mechanism as well so it's very likely that the issue is either related to some request payload (headers or body) or the proxy in use.

**Comment by Dangerunicorn** [comment]
I have no request body in the one I'm using to test. (I have been checking other tests as well to be sure.) The request I'm using is a basic call to our website to ask which user is returned with the credentials given. 

I noticed that I have the same Headers as were mentioned in #4355. I see these in the console log. 
WWW-Authenticate: Negotiate
WWW-Authenticate: NTLM
X-Powered-By: ASP.NET

**Comment by codenirvana** [comment]
@Dangerunicorn Not sure its the exact same problem. Because in your case I see the last header is `NTLM` not `Negotiate` and since the current implementation always choose the last header (if duplicates) the request should work.

Can you clear the console and send the request once and for evert request log expand the request and response header section (hide sensitive details) and share the screenshot?

**Comment by Dangerunicorn** [comment]
Below you will see my console log. I am only sending one request called "Who Am I?". There is nothing in the body. I am only sending the request once. All of my requests appear this way - running 3 times. They only succeed on the third try. 

![1](https://user-images.githubusercontent.com/54122108/65899109-e551f100-e367-11e9-89fc-1a1b3dc938dd.PNG)
![2](https://user-images.githubusercontent.com/54122108/65899113-e8e57800-e367-11e9-930f-b8cef151d759.PNG)
![3](https://user-images.githubusercontent.com/54122108/65899116-eaaf3b80-e367-11e9-9ab1-157a12578ed8.PNG)

**Comment by codenirvana** [comment]
> They only succeed on the third try.

Do you mean the third request is `200 OK` or you have to send the request 3 times to make it work?

> All of my requests appear this way - running 3 times

Yes, these are the NTLM handshake requests.

From the request-response screenshots looks like server rejected the type 3 message (third request) which I think is because of invalid credentials or server error. 

As you can see it's not related to #4355 as the type 2 message (second request's response) doesn't have multiple `www-authenticate` response headers.

**Comment by Dangerunicorn** [comment]
I don't send the request 3 times. I send the request once and postman sends it three times. Every response is 401 as you can see in the screen shots above.

These credentials are not invalid. I can access this end point in browser manually with no issue. I can also access the site in incognito mode, and I can enter in the same credentials without issue. 

How can I tell if this is a server error?

**Comment by codenirvana** [comment]
Yes as previously said those three requests sent are part of NTLM handshake (no issues with that).

> How can I tell if this is a server error?

Not sure, so far I am not able to understand what's the issue with this request.

If possible can you share a sample endpoint and required credentials with us @ `help@getpostman.com` so that we can reproduce this issue internally.

**Comment by Dangerunicorn** [comment]
Unfortunately there is no way we can provide a sample endpoint or credentials. I wish I could send that info your way, but everything I have is internal for our company and touches sensitive user information. 

Is there any other way I could possibly assist in helping you to narrow this down? Is there a public site API available with NTLM authentication that I can test against? Maybe then I could give you all the information as I see it.

**Comment by Dangerunicorn** [comment]
Is there any possible way that we can proceed? This is holding up work for us, and I don't know where to turn. 

Can we unblock this issue? It is an ongoing problem that has halted work for us.

**Comment by Dudak22** [comment]
+1

**Comment by saswatds** [comment]
Hi @Dangerunicorn, 
As you maybe aware, NTML is a proprietary protocol designed by Microsoft with no publicly available specification. Over the time frame you have mentioned we have had made no changes to NTML implementation in our runtime. The working of the NTLM(beta) auth feature greatly depends on how the IIS server has been configured on your end. Without the completely knowledge of your configuration it's very difficult for us to know why things are not working.
Following one of the official blogs of Microsoft
https://blogs.msdn.microsoft.com/chiranth/2013/09/20/ntlm-want-to-know-how-it-works/

For NTLM to work 3 request have to be made, and that is what you see in the console.
* Request 1 is made where the IIS server will respond with 401 and www-authenticate header requesting for NTML to be used
* Request 2 is made to the server with Authorization header set to NTML with domain and workstation information, For with the server responds with a challenge in www-authenticate header
* Request 3 is made with the Authorization header set to a type3 message with the username and password.

As you can see that all the three request were correctly made but the server still returned a 401 for final request.

To put us in a right direction for resolution, It would help if you can try out the request you are making with this node library https://github.com/SamDecrock/node-http-ntlm . The readme of the library should get you started. This is the library we use internally to compute and parse the different message. If you are able to successfully make the request work, let us know the configuration that worked for you so that we can debug.

Also I am bit confused on the state of the issue.  Does NTML not work at all for your *right now* or does it work but fails intermittently (works after like 2-3 tries)?
Also check for any whitespace character in the username/password fields that could have creeped in causing the request to fail.

**Comment by Dangerunicorn** [comment]
3 Responses in Console: Totally understand where that's coming from. The example provided above was one GET request sent once, I just wanted to be sure to include the 3 times that it runs with all the info. 

Credentials: My credentials are stored in environment variables. These haven't changed and I've double checked for you to confirm they are accurate. These same credentials get me to the website fine in browser. I have another couple of users credentials that I'm using to test as well. I will be using all 3 users to test to be totally sure I'm seeing the same results across the board. 

State of Issue: NTLM does not work at all for me at this time. We have had other issues with NTLM in the past and are currently using a work around. NTLM Authentication does not inherit from parent, so we've had to manually put the credentials in every test. We use environment variables for this so I can change the user quickly without any possible user error for mistyping in credentials. Again that's a separate issue, but I wanted you to be aware that NTLM never worked perfectly for us. 

Node HTTP NTLM: I've passed this solution to one of the developers on our team to see if he can get this to work. He said at this time he's been unsuccessful but he's going to keep working at it. I'll let you know if we have any luck. 

Additional Suggestions: Do you have any additional suggestions of where to go from here? Where we can look for more information? Or even suggestions on how to configure this Node HTTP NTLM to get it to work? This problem has been a baffling one for us.

**Comment by saswatds** [comment]
NTML is inherently very tricky. The information that would help the most would be to know from the server why request's from postman are failing, but I presume it will be difficult.

I can see that you are using a proxy so the following snippet should work
```
var httpntlm = require('httpntlm');

httpntlm.get({
    url: '<url>',
    username: '<username>',
    password: '<password>',
    workstation: '', // Optionally fill if you already do
    domain: '',  // Optionally fill if you already do
    proxy: {
     host: '<host>',
     port: '<port>'
    }
}, function (err, res){
    if(err) return err;

    console.log(res.headers);
    console.log(res.body);
});
```

With regards to the credentials. when you are using environment variables it is still possible that spaces creep in the following manner. If you can just make sure the `{{variable}}` is not surrounded with any kind of space would be help. 
<img width="536" alt="Screenshot 2019-10-02 at 12 28 07 AM" src="https://user-images.githubusercontent.com/8796447/65991976-e8ee9080-e4ab-11e9-9b37-9a58652c2e83.png">

**Comment by Dangerunicorn** [comment]
Thank you for the snippet. I'll pass that along to our developer as well.

I have confirmed that there are no spaces before or after the username, password, or domain.

**Comment by kirk911** [comment]
I could not get NTLM to work using Postman.  I tried the exact same request using Curl and the --ntlm flag and it worked without issue.  Perhaps you could try with Curl to rule out an issue with your network?

**Comment by Dangerunicorn** [comment]
Sorry for the few day hiatus. We've discovered the root of our issue but not the cause. Something has changed with our internal authentication and the domain is no longer required. I'm not sure why the request breaks when domain is provided (it never did before). I still need to contact our cooperate IT department to find out what exactly changed. Nothing else has broken or changed with this "update". 

If anyone else finds themselves in a similar issue, I urge you to contact your IT to ask if anything about your cooperate authentication changed. There's a chance they changed things and didn't alert the company, as it was in my case. 

Postman... it's not you... it's me.
